lib/libvfio-user.c

```diff
@@ -1976,19 +2083,71 @@ vfu_sg_is_mappable(vfu_ctx_t *vfu_ctx, dma_sg_t *sg)
     return dma_sg_is_mappable(vfu_ctx->dma, sg);
 }
 
-EXPORT void
-vfu_migr_done(vfu_ctx_t *vfu_ctx, int reply_errno)
+EXPORT int
+vfu_device_quiesced(vfu_ctx_t *vfu_ctx, int quiesce_errno)
 {
+    int ret;
+
     assert(vfu_ctx != NULL);
-    assert(vfu_ctx->migr_trans_pending);
 
-    if (!vfu_ctx->migr_trans_msg->hdr.flags.no_reply) {
-        do_reply(vfu_ctx, vfu_ctx->migr_trans_msg, reply_errno);
+    if (vfu_ctx->quiesce == NULL
+        || vfu_ctx->pending.state == VFU_CTX_PENDING_NONE) {
+        vfu_log(vfu_ctx, LOG_DEBUG,
+                "invalid call to quiesce callback, state=%d",
+                vfu_ctx->pending.state);
+        return ERROR_INT(EINVAL);
     }
-    free_msg(vfu_ctx, vfu_ctx->migr_trans_msg);
-    vfu_ctx->migr_trans_msg = NULL;
 
-    vfu_ctx->migr_trans_pending = false;
+    vfu_log(vfu_ctx, LOG_DEBUG, "device quiesced with error=%d", quiesce_errno);
+
+    if (quiesce_errno == 0) {
+        switch (vfu_ctx->pending.state) {
+        case VFU_CTX_PENDING_MIGR:
+            ret = migr_trans_to_valid_state(vfu_ctx, vfu_ctx->migration,
+                                            vfu_ctx->pending.migr_dev_state, true);
+            break;
+        case VFU_CTX_PENDING_DMA_MAP:
```

Comment:
Why not this:

```
 case VFU_CTX_PENDING_MSG:
    ret = exec_command(vfu_ctx, vfu_ctx.pending.msg);
    break;
```

and drop the map/unmap/device_reset_quiesced() variants altogether, along with .device_reset_reason. All you'd need to change would be don't call ->quiesce() when there's already ->pending.state (although I think I'd prefer an explicit "quiesced" state tracking anyway)

In fact, given my above point about FLR, I'm wondering - can we do this for *all* region access too. Instead of tracking sub-state for the region write (FLR or migration device state), peek ahead first:

```
 198 static ssize_t                                                                   
 199 region_access(vfu_ctx_t *vfu_ctx, size_t region_index, char *buf,                
 200               size_t count, uint64_t offset, bool is_write)                      
 201 {                                                                                

...
       if (access_needs_quiesce()) {
           ... call quiesce, handle EBUSY if needed ...
       }
...
```

which would check for the migration device state offset or the PCI FLR offset. I think this works out simpler.

Reply:
The idea is to first check whether quiesce is required and if so quiesce the device and thern re-run `exec_command` once the device is quiesced.

Reply:
@jlevon checking for the FLR offset isn't trivial as it's the relevant PCI capability can be anywhere past 0x40 in the PCI config space. The simplest way would be to call `cap_find_by_offset` and see whether a capability is returned and if it's the PCI Express capability then check whether it's FLR being written. This will make any PCI config space access a bit slower. Another way would be to store the offset of the PCI express capability in `vfu_ctx` so that we can easily look it up. This will, however, result in keeping redundant information that can be harder to maintain. Thoughts?

Reply:
Agreed, let's cache it

Reply:
It just occured to me that if the user handles entirely the PCI config space then we won't know that relevant FLR register is written to so we won't tell the device to quiesce. One idea is to look for this capability in the PCI config space when `vfu_realize_ctx` is called. Otherwise we could just not handle this at all and leave it to the user.

Reply:
good point, let's file a ticket on that for now, and check with Ram later.

Reply:
https://github.com/nutanix/libvfio-user/issues/629